### PR TITLE
Rework rustdoc const type

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
This PR is mostly about two things:
 1. Not storing some information in the `clean::Constant` type
 2. Using `TyCtxt` in the formatting (which we will need in any case as we move forward in any case).

Also: I'm very curious of the perf change in here.

Thanks a lot @danielhenrymantilla for your `Captures` idea! It allowed me to solve the lifetime issue completely. :)

r? @jyn514

### CODE REVIEW

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -1745,12 +1743,7 @@ impl<'tcx> Clean<Type> for Ty<'tcx> {
 
 impl<'tcx> Clean<Constant> for ty::Const<'tcx> {
     fn clean(&self, cx: &mut DocContext<'_>) -> Constant {
-        Constant {
-            type_: self.ty.clean(cx),
-            expr: format!("{}", self),
-            value: None,
-            is_literal: false,
-        }
+        Constant::TyConst { type_: self.ty.clean(cx), expr: format!("{}", self) }
```
Comment (jyn514) [inline src/librustdoc/clean/mod.rs:1746]: Rather than stringifying `self`, could you store it in the `Constant` and only stringify on demand? Or does that make the lifetimes painful?

File: src/librustdoc/clean/types.rs [hunk 2]
```diff
@@ -1966,11 +1967,58 @@ crate struct Static {
...
-    crate type_: Type,
-    crate expr: String,
-    crate value: Option<String>,
-    crate is_literal: bool,
+crate enum Constant {
+    TyConst { type_: Type, expr: String },
+    Generic { type_: Type, body: BodyId },
+    Inline { type_: Type, did: DefId },
+    Const { type_: Type, did: DefId, body: BodyId },
+}
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1975]: Can you document what these are? I can guess `Inline` is for `const { ... }` blocks, but I can't tell about the rest.

File: src/librustdoc/html/render/mod.rs [hunk 3]
```diff
@@ -1052,11 +1053,10 @@ fn render_assoc_item(
...
             name = name,
-            generics = g.print(cx.cache()),
-            decl = Function { decl: d, header_len, indent, asyncness: header.asyncness }
-                .print(cx.cache()),
-            spotlight = spotlight_decl(&d, cx.cache()),
-            where_clause = WhereClause { gens: g, indent, end_newline }.print(cx.cache())
+            generics = g.print(cx.cache(), cx.tcx()),
+            decl = d.full_print(cx.cache(), cx.tcx(), header_len, indent, header.asyncness),
+            spotlight = spotlight_decl(&d, cx.cache(), cx.tcx()),
+            where_clause = print_where_clause(g, cx.cache(), cx.tcx(), indent, end_newline),
```
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:1059]: You use `tcx()` quite a lot - maybe assign it to a temporary?

Note that this will have pretty big conflicts with https://github.com/rust-lang/rust/pull/82855.

**Review by jyn514 — COMMENTED** [review]
I haven't read the whole thing yet. Could you describe more the motivation for this change? You said

> Not storing some information in the clean::Constant type

which makes sense to me - is the idea just to help performance?

File: src/librustdoc/clean/mod.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs:1746]: It's worth giving it a try.

File: src/librustdoc/clean/types.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1975]: Arf sorry, I was experimenting on it for so long that I didn't document things...

File: src/librustdoc/html/render/mod.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:1059]: I'll rebase then. ;)

**Comment by GuillaumeGomez** [comment]
The benchmark results seem mostly positive (just wondering what's happening with `ctfe-stress-4-doc`):

Screenshot from 2021-03-07 22-37-53

**Comment by GuillaumeGomez** [comment]
> I haven't read the whole thing yet. Could you describe more the motivation for this change? You said
> 
> > Not storing some information in the clean::Constant type
> 
> which makes sense to me - is the idea just to help performance?

It's definitely targetting performance here. It's also preparing the field in `html/format` by passing the `TyCtxt` type around since we'll very likely make more things into the formatting directly too (and this is where most changes actually come from...).

File: src/librustdoc/clean/mod.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/mod.rs:1746]: So, after taking a quick look, I propose this instead: I'll add a FIXME so we did it in another PR because this will otherwise just grow the diff once again, and it's already huge.

**Comment by GuillaumeGomez** [comment]
So let's see if your idea will improve the perf, even a little. ;)

@bors try @rust-timer queue

**Comment by jyn514** [comment]
> It's also preparing the field in html/format by passing the TyCtxt type around since we'll very likely make more things into the formatting directly too (and this is where most changes actually come from...).

Ok, so the TyCtxt isn't actually needed for these changes? Could you split it into a different PR (or just commit) so it's easier to see which changes are necessary?

**Comment by GuillaumeGomez** [comment]
Well... Actually it is needed because you need it to print the const. And from that, all the "upper" functions needed it too. You'll note that not all functions in `html/format` take a `TyCtxt`, that's the reason. I "only" changed the ones which needed it because one of the children was using it. Pretty impressive how one change is affecting so many things. If it wasn't needed yet, I wouldn't have been able to add a parameter without using underscore to prevent warnings. ;)

**Comment by GuillaumeGomez** [comment]
No notable changes since before. Waiting for the review then!

File: src/librustdoc/clean/types.rs [hunk 4]
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
 }
 
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
-crate struct Constant {
-    crate type_: Type,
-    crate expr: String,
-    crate value: Option<String>,
-    crate is_literal: bool,
+crate enum Constant {
+    /// Typed constant value.
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1971]: What does this mean? Aren't all values typed?

File: src/librustdoc/clean/types.rs [hunk 5]
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
...
+    TyConst { type_: Type, expr: String },
+    /// A constant (expression) that’s not an item or associated item. These are usually found
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Generic { type_: Type, body: BodyId },
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1976]: Generic seems like a bad name for this. Maybe "Anonymous" is better?

File: src/librustdoc/clean/types.rs [hunk 6]
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
...
+    /// Inlined constant (from another crate).
+    Inline { type_: Type, did: DefId },
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1978]: Constants can be inlined without being from another crate. I think `Extern` or `ExternItem` would be a better name.

File: src/librustdoc/clean/types.rs [hunk 7]
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
...
+    /// const FOO: u32 = ...;
+    Const { type_: Type, did: DefId, body: BodyId },
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1980]: "Const" is not very descriptive. Maybe "Local" or "Item"?

**Review by jyn514 — COMMENTED** [review]
> The benchmark results seem mostly positive (just wondering what's happening with `ctfe-stress-4-doc`):

Wow, that's amazing! I looked at the self-profile data and it looks like we've almost completely eliminated the time `eval_to_allocation_raw` takes to execute (!!). Do you know how that happened? I don't see it being called in rustdoc directly.

> eval_to_allocation_raw | 0.002 | 0.39% | -99.9% (-1.191) | 151 | -5.0% (-8.000)

I haven't looked at the rest of the changes in detail but this definitely seems worth landing if it has that big an impact :)

**Comment by GuillaumeGomez** [comment]
> > The benchmark results seem mostly positive (just wondering what's happening with `ctfe-stress-4-doc`):
> 
> Wow, that's amazing! I looked at the self-profile data and it looks like we've almost completely eliminated the time `eval_to_allocation_raw` takes to execute (!!). Do you know how that happened? I don't see it being called in rustdoc directly.

I assume that it was called indirectly, as most things. But to be honest, I didn't expect this to have such a great impact. :)

> > eval_to_allocation_raw | 0.002 | 0.39% | -99.9% (-1.191) | 151 | -5.0% (-8.000)
> 
> I haven't looked at the rest of the changes in detail but this definitely seems worth landing if it has that big an impact :)

Awesome! Fixing your comments then as quickly as possible then!

File: src/librustdoc/clean/types.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1971]: I took the documentation from https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Const.html

If you have ideas for improvements, it's very welcome. :laughing:

File: src/librustdoc/clean/types.rs (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1976]: The original type is [AnonConst](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.AnonConst.html) so you're right, it'd be much better!

File: src/librustdoc/clean/types.rs (hunk 6 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1978]: Let's go for `Extern` then.

File: src/librustdoc/clean/types.rs (hunk 7 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1980]: Let's go for `Local` then!

**Comment by GuillaumeGomez** [comment]
And conflicts fixed!

**Comment by GuillaumeGomez** [comment]
Fixed format issue.

File: src/librustdoc/clean/types.rs (hunk 4 above)
Comment (camelid) [reply src/librustdoc/clean/types.rs:1971]: I \*think\* it might be referring to the fact that [`ConstValue`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/interpret/value/enum.ConstValue.html) is basically loosely-structured raw bytes and you need the type to figure out what it actually means.

Comment (jyn514) [reply src/librustdoc/clean/types.rs:1971]: > I took the documentation from https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Const.html

Ah ok, you literally mean `ty::Const`, I thought it had some deeper meaning. And the reason it can't be combined with `Local` is because non-local constants won't have a `BodyId`. Can you add both those things as a comment? And please remove "A typed constant value" - that would be good documentation for `enum Constant`, but it's not helpful for this variant because it's true for *all* the `Constant` variants.

File: src/librustdoc/clean/mod.rs [hunk 8]
```diff
@@ -1135,7 +1133,7 @@ impl Clean<Item> for ty::AssocItem {
             ty::AssocKind::Const => {
                 let ty = tcx.type_of(self.def_id);
                 let default = if self.defaultness.has_value() {
-                    Some(inline::print_inlined_const(cx, self.def_id))
+                    Some(inline::print_inlined_const(cx.tcx, self.def_id))
```
Comment (jyn514) [inline src/librustdoc/clean/mod.rs:1136]:
```suggestion
                    Some(inline::print_inlined_const(tcx, self.def_id))
```

File: src/librustdoc/clean/types.rs [hunk 9]
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
...
+    }
+
+    crate fn to_type(self) -> Type {
+        match self {
+            Self::TyConst { type_, .. }
+            | Self::Extern { type_, .. }
+            | Self::Local { type_, .. }
+            | Self::Anonymous { type_, .. } => type_,
+        }
+    }
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:2027]: Have you considered using a struct that holds the type instead? Then you wouldn't need these two functions and you could just access the type directly:
```rust
struct Constant {
  type_: Type,
  kind: ConstantKind,
}

enum ConstantKind { ... }
```

File: src/librustdoc/html/format.rs [hunk 10]
```diff
@@ -107,31 +107,6 @@ impl Buffer {
...
-
-/// Wrapper struct for emitting a where-clause from Generics.
-crate struct WhereClause<'a> {
-    /// The Generics from which to emit a where-clause.
-    crate gens: &'a clean::Generics,
-    /// The number of spaces to indent each line with.
-    crate indent: usize,
-    /// Whether the where-clause needs to add a comma and newline after the last bound.
-    crate end_newline: bool,
-}
```
Comment (jyn514) [inline src/librustdoc/html/format.rs:133]: Where did these go?

File: src/librustdoc/html/format.rs [hunk 11]
```diff
@@ -966,147 +1012,176 @@ impl clean::BareFunctionDecl {
...
+    /// * `asyncness`: Whether the function is async or not.
+    crate fn full_print<'b, 'a: 'b, 'tcx: 'a>(
+        &'a self,
+        cache: &'b Cache,
+        tcx: TyCtxt<'tcx>,
+        header_len: usize,
+        indent: usize,
+        asyncness: hir::IsAsync,
+    ) -> impl fmt::Display + 'b + Captures<'tcx> {
+        display_fn(move |f| self.inner_full_print(cache, tcx, header_len, indent, asyncness, f))
```
Comment (jyn514) [inline src/librustdoc/html/format.rs:1056]: Why is this new function needed?

File: src/librustdoc/html/render/mod.rs [hunk 12]
```diff
@@ -1172,21 +1181,19 @@ fn render_assoc_items(
             AssocItemRender::DerefFor { trait_, type_, deref_mut_ } => {
                 let id = cx.derive_id(small_url_encode(format!(
                     "deref-methods-{:#}",
-                    type_.print(cx.cache())
+                    type_.print(cache, tcx)
```
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:1184]: Are all these changes from `cx.cache()` to `cache` necessary? They're not *bad*, but they make an already giant diff even bigger, it would be nice to put them in a separate commit or something.

File: src/librustdoc/html/render/mod.rs [hunk 13]
```diff
@@ -1582,7 +1595,16 @@ fn render_impl(
...
+                assoc_type(
+                    w,
+                    item,
+                    bounds,
+                    default.as_ref(),
+                    link.anchor(&id),
+                    "",
+                    cx.cache(),
+                    tcx,
+                );
```
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:1607]: Rather than passing both of these, could you just pass a `cx`?

Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:1607]: Rather than passing both of these, could you just pass a `cx`?

File: src/librustdoc/clean/types.rs [hunk 14]
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
...
+            Self::Inline { did, .. } | Self::Const { did, .. } => print_evaluated_const(tcx, *did),
+        }
+    }
+
+    crate fn is_literal(&self, tcx: TyCtxt<'_>) -> bool {
+        match self {
+            Self::TyConst { .. } => false,
+            Self::Inline { did, .. } => did
+                .as_local()
+                .map_or(false, |did| is_literal_expr(tcx, tcx.hir().local_def_id_to_hir_id(did))),
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs]: I took this code from here: https://github.com/rust-lang/rust/blob/4d7a6484852603749f765479ee7288e6ae7186f4/src/librustdoc/clean/inline.rs#L507-L509

I didn't change its behavior.

File: src/librustdoc/clean/types.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs]: Sure!

File: src/librustdoc/clean/types.rs [hunk 15]
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
...
-    crate is_literal: bool,
+crate enum Constant {
+    /// Typed constant value.
+    TyConst { type_: Type, expr: String },
+    /// A constant (expression) that’s not an item or associated item. These are usually found
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Anonymous { type_: Type, body: BodyId },
+    /// Inlined constant.
+    Extern { type_: Type, did: DefId },
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1978]: This one uses `print_inlined_const` and is generated from `clean/inline.rs` whereas `TyConst` simply uses `to_string()` from its expression.

File: src/librustdoc/clean/types.rs (hunk 9 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:2027]: I did but I wondered for a while if it was worth it or not. I personally think that the current code is fine, but if you prefer otherwise, I can update.

File: src/librustdoc/html/format.rs (hunk 10 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/format.rs:133]: Removed and replaced by functions.

File: src/librustdoc/html/render/mod.rs (hunk 13 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:1607]: Same: I can, but then I'd have to call again both `cx.cache()` and `cx.tcx()`. So as you prefer?

File: src/librustdoc/html/format.rs (hunk 11 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/format.rs:1056]: To be more precise: there was a huge lifetime issue so I fixed it by simply removing the two structs.

File: src/librustdoc/html/format.rs (hunk 10 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/format.rs:133]: To be more precise: there was a huge lifetime issue so I fixed it by simply removing the two structs.

File: src/librustdoc/clean/types.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/types.rs]: Hmm, ok - it doesn't need to change here then.

File: src/librustdoc/clean/mod.rs [hunk 16]
```diff
@@ -1745,12 +1743,8 @@ impl<'tcx> Clean<Type> for Ty<'tcx> {
 
 impl<'tcx> Clean<Constant> for ty::Const<'tcx> {
     fn clean(&self, cx: &mut DocContext<'_>) -> Constant {
-        Constant {
-            type_: self.ty.clean(cx),
-            expr: format!("{}", self),
-            value: None,
-            is_literal: false,
-        }
+        // FIXME: instead of storing `format!("{}", self)`, store `self` directly instead.
```
Comment (jyn514) [inline src/librustdoc/clean/mod.rs:1746]:
```suggestion
        // FIXME: instead of storing the stringified expression, store `self` directly instead.
```

File: src/librustdoc/clean/types.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/types.rs]: Right, sure, but why do you need it to be separate? Could you store this instead wherever you store a `TyConst`?

File: src/librustdoc/clean/types.rs (hunk 9 above)
Comment (jyn514) [reply src/librustdoc/clean/types.rs:2030]: I think I would rather have the explicit field, yeah.

File: src/librustdoc/html/render/mod.rs (hunk 13 above)
Comment (jyn514) [reply src/librustdoc/html/render/mod.rs:1617]: Nah, this is fine.

File: src/librustdoc/clean/types.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs]: Because it's a reference and adding a lifetime everywhere seemed like a bit too much for this PR. :laughing: 

Do you want me to add a FIXME so we can do it in another PR?

File: src/librustdoc/clean/types.rs [hunk 17]
```diff
@@ -1966,11 +1967,67 @@ crate struct Static {
...
+crate enum Constant {
+    /// This is the wrapper around `ty::Const` for a non-local constant. Because it doesn't have a
+    /// `BodyId`, we need to handle it on its own.
+    TyConst { type_: Type, expr: String },
+    /// A constant (expression) that’s not an item or associated item. These are usually found
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Anonymous { type_: Type, body: BodyId },
+    /// Inlined constant.
+    Extern { type_: Type, def_id: DefId },
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1979]: This is specifically for non-local constants I thought? Or does `clean::inline` discard the BodyId? That doesn't seem right.
```suggestion
    /// A constant from a different crate.
    Extern { type_: Type, def_id: DefId },
```

File: src/librustdoc/clean/types.rs [hunk 18]
```diff
@@ -1966,11 +1967,67 @@ crate struct Static {
...
+impl Constant {
+    crate fn expr(&self, tcx: TyCtxt<'_>) -> String {
+        match self {
+            Self::TyConst { expr, .. } => expr.clone(),
+            Self::Extern { def_id, .. } => print_inlined_const(tcx, *def_id),
+            Self::Local { body, .. } | Self::Anonymous { body, .. } => print_const_expr(tcx, *body),
+        }
+    }
+
+    crate fn value(&self, tcx: TyCtxt<'_>) -> Option<String> {
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1993]: What does 'value' mean? Why do only `Extern` and `Local` constants have one?

File: src/librustdoc/clean/types.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/types.rs]: > Because it's a reference and adding a lifetime everywhere seemed like a bit too much for this PR. laughing

Can't you just store the `DefId` instead of storing the whole `ty::Const`?

File: src/librustdoc/clean/types.rs (hunk 9 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:2030]: Then let's go for it!

File: src/librustdoc/clean/types.rs (hunk 18 above)
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1993]: It's the `= "whatever";` part. I took the "value" name from the field of the original `clean::Constant` struct.

**Comment by GuillaumeGomez** [comment]
Split the enum into a struct and an enum. :)

**Comment by GuillaumeGomez** [comment]
@camelid Updated the documentation. :)

File: src/librustdoc/clean/types.rs (hunk 18 above)
Comment (jyn514) [reply src/librustdoc/clean/types.rs:2001]: How is it different from 'expr'?

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:2001]: That's actually a good question now that I think about it. :laughing:

Comment (jyn514) [reply src/librustdoc/clean/types.rs:2001]: Can you try switching all uses of `value()` to use `expr()` instead and see if it breaks anything? If not, could you remove this function?

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:2001]: It's worth a try.

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:2001]: So I think it's worth thinking about it later on because:

```rust
let value = c.value(cx.tcx());
let is_literal = c.is_literal(cx.tcx());
let expr = c.expr(cx.tcx());
if value.is_some() || is_literal {
    write!(w, " = {expr};", expr = Escape(&expr));
} else {
    w.write_str(";");
}

if !is_literal {
    if let Some(value) = &value {
        let value_lowercase = value.to_lowercase();
        let expr_lowercase = expr.to_lowercase();

        if value_lowercase != expr_lowercase
            && value_lowercase.trim_end_matches("i32") != expr_lowercase
        {
            write!(w, " // {value}", value = Escape(value));
        }
    }
}
```

Both `expr` and `value` are used.

Comment (jyn514) [reply src/librustdoc/clean/types.rs:2001]: That doesn't really answer my question though  - why not just unconditionally print the expression, instead of printing it for some types of constants not others?

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:2001]: It needs to be investigated. But I don't think it's a good idea to do it here. Should I open an issue about this?

Comment (jyn514) [reply src/librustdoc/clean/types.rs:2001]: Hmm, ok, we can look into it in a follow-up. If you open an issue that would be great.

**Comment by jyn514** [comment]
This is waiting on https://github.com/rust-lang/rust/pull/82873#discussion_r591818612, and then I need to read the full diff.

File: src/librustdoc/clean/types.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs]: The `ty::Const` comes from https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/subst/enum.GenericArgKind.html so I wonder if `tcx.type_of(def_id)` will actually allow me to get it without too much troubles.

The easiest way seem to simply remove the clean impl for `ty::Const` and instead take it by value into the enum. However, it has a lifetime which is what we don't want to do here if we want to avoid a super big diff. I already added a FIXME, so unless there is a way to get the `ty::Const` from the `DefId`, I think it's better to keep it this way here and then move the `ty::Const` directly into the enum in another PR (and with all the lifetime changes it'll bring).

However, if you know a way to get the `ty::Const`, I'd be more than happy to do it here directly. I just couldn't find a way to get it though...

File: src/librustdoc/clean/types.rs [hunk 19]
```diff
@@ -1968,9 +1969,55 @@ crate struct Static {
...
+crate enum ConstantKind {
+    /// This is the wrapper around `ty::Const` for a non-local constant. Because it doesn't have a
+    /// `BodyId`, we need to handle it on its own.
+    TyConst { expr: String },
+    /// A constant (expression) that's not an item or associated item. These are usually found
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Anonymous { body: BodyId },
+    /// A constant from a different crate.
+    Extern { def_id: DefId },
```
Comment (camelid) [inline src/librustdoc/clean/types.rs:1985]: I feel confused about this: `TyConst` says (emphasis mine)

> This is the wrapper around `ty::Const` for a **non-local constant**.

but `Extern` *also* says

> A constant from a **different crate**.

What's the difference between a "non-local constant" and a "constant from a different crate"?

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1985]: `TyConst` is about generics (meaning, "non local") whereas `Extern` is for foreign constants items.

Comment (camelid) [reply src/librustdoc/clean/types.rs:1985]: When you say "about generics" do you mean it's about const generics? What's an example of a `TyConst` and of an `Extern`?

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1985]: You can only find `TyConst` when listing over an item generic arguments. An extern is simply a const variable defined in another crate.

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1985]: Take a look at `fn external_generic_args` in `clean/utils.rs`.

Comment (camelid) [reply src/librustdoc/clean/types.rs:1985]: Why do we need to track `Extern` separately from `Local`? It seems like you can get the value of the `Extern` just from the `DefId`, so can't we merge `Extern` and `Local`?

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1985]: The way they're displayed is both different (one use a `BodyId` whereas the other is forced to use its `DefId`. But ultimately, it'd be nice to be able to merge them. But I think it's out of the scope of this PR.

**Comment by GuillaumeGomez** [comment]
@camelid @jyn514 Anything else to be done here? (Kinda worried about the coming rebases ^^')

**Comment by camelid** [comment]
> (Kinda worried about the coming rebases ^^')

(What rebases?)

**Comment by GuillaumeGomez** [comment]
Development in rustdoc is still going on so conflicts will very likely appear. Which is why I'd prefer to not keep this PR hanging around for too long.

File: src/librustdoc/clean/types.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/types.rs]: @GuillaumeGomez why did you mark this as resolved? I vaguely remember we talked about this in DMs but I don't remember what we decided - I still feel like TyConst could be combined into `Extern` by only storing the DefId.

[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/types.rs]: I guess my question more broadly is why you need the `ty::Const` for `TyConst` when you don't need it for `Extern`.

**Comment by jyn514** [comment]
> Development in rustdoc is still going on so conflicts will very likely appear. Which is why I'd prefer to not keep this PR hanging around for too long.

I'm sorry, I can't commit to reviewing in any period of time. If you'd like it to be reviewed sooner you could maybe ping some other team members, but they may have other commitments too.

File: src/librustdoc/clean/types.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs]: Because `ty::Const` is a generic, not a type and can't be retrived easily just from its `DefId`. Which is why you said on zulip that it was fine to keep as is.

[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/types.rs]: Thanks! That was it :)

File: src/librustdoc/json/conversions.rs [hunk 20]
```diff
@@ -87,59 +87,78 @@ impl JsonRenderer<'_> {
...
+}
+
+impl<T, U> IntoWithTcx<U> for T
+where
+    U: FromWithTcx<T>,
+{
+    fn into_tcx(self, tcx: TyCtxt<'_>) -> U {
+        U::from_tcx(self, tcx)
+    }
+}
```
Comment (jyn514) [inline src/librustdoc/json/conversions.rs:105]: Why do these need to be traits? It seems simpler to just use free functions. Otherwise you'll have to redo this all again next time you want to add a parameter.

**Review by jyn514 — COMMENTED** [review]
I think I'm ready to approve this with a couple nits fixed. I don't feel confident I've caught all the bugs, but hopefully the test suite will pick up anything that goes wrong.

File: src/librustdoc/clean/types.rs [hunk 21]
```diff
@@ -1978,9 +1979,55 @@ crate struct Static {
...
-    crate value: Option<String>,
-    crate is_literal: bool,
+    crate kind: ConstantKind,
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+crate enum ConstantKind {
+    /// This is the wrapper around `ty::Const` for a non-local constant. Because it doesn't have a
+    /// `BodyId`, we need to handle it on its own.
+    TyConst { expr: String },
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1989]: :+1:

File: src/librustdoc/json/conversions.rs (hunk 20 above)
Comment (GuillaumeGomez) [reply src/librustdoc/json/conversions.rs:105]: I kept it as close as possible to the current code (which was using `From`). Otherwise I'll have to replace all the current trait implementations with free functions for all different types.

**Comment by GuillaumeGomez** [comment]
Rebased to fix conflicts and updated doc comment as suggested by @jyn514. Now remains https://github.com/rust-lang/rust/pull/82873#discussion_r600151523.

File: src/librustdoc/json/conversions.rs (hunk 20 above)
Comment (jyn514) [reply src/librustdoc/json/conversions.rs:105]: Hmm, well I'm not convinced the traits are pulling their weight, but we can fix it in a follow-up to avoid unnecessary churn.
